### ISSUE TITLE
More control to vnode equality comparison during patch

### ISSUE THREAD

**Issue by milankinen** [paldepind/snabbdom#95]
Hi and thanks for the great library! I noticed this issue when I was implementing `snabbdom` interpreter to [`TSERS`](https://github.com/tsers-js). 

Reactive libs (e.g. RxJs, Kefir, ...) usually have a combinator function for arrays of observables (like `Rx.Observable.combineLatest`) that cache the latest values from the combined observables and return the cached values from other observables than the changed one. 

I discovered that if I remove one item from the existing observable list (`list$ = O.combine(item1$, item2$, ...., render)`) the list rendering crashes. It seems that happens because `snabbdom` *mutates*  the patched/rendered new vnode by adding `elm` into it. However, because the items are cached inside the `O.combine`, the next patch/render (after item removal) receives *exactly same vnodes* as old and new ones. If I remove one item from the beginning of the list, `snabbdom` starts to patch and mutate existing vnodes, hence causing unexpected behaviour (somethings it crashes, sometimes just renders the DOM incorrectly, depending on the content).

Here is the example that is broken:
```js
var snabbdom = require('snabbdom');
var patch = snabbdom.init([ ]);

var h = require('snabbdom/h'); 

var root = document.getElementById('app')

var li_1 = [
  h("li", [h("span", ["1"])]),
  h("li", [h("span", ["2"])]),
  h("li", [h("span", ["3"])]),
  h("li", [h("span", ["4"])])
]

var first = h('div', [
  h("ul", li_1)
])


var li_2 = [
  li_1[0],
  li_1[2],
  li_1[1],
  li_1[3],
]

var second = h('div', [
  h("ul", li_2)
])

patch(root, first)
patch(first, second)
```
In the end, that should render `1, 3, 2, 4`. However, the final output is `1,2,3,4`.

Fortunately this can easily be avoided by cloning the vtree before the patching operation:
```js
function cloneVTree(vnode) {
  return {
    sel: vnode.sel,
    data: vnode.data,
    children: vnode.children && vnode.children.map(cloneTree),
    text: vnode.text,
    key: vnode.key,
    elm: vnode.elm
  }
}
```  
However, this disables a very important performance optimization: [vnode identity check](https://github.com/paldepind/snabbdom/blob/fb80c1fd80235e93803d51a7ab274d659fd1a317/snabbdom.js#L204).

My proposal is to give **more control to vnode's equality comparison**. The solution would be extremely simple: optional `.equals` function to vnode which perform the equality comparison. The solution would require only one backwards compatible change:
```diff
    ...
    var elm = vnode.elm = oldVnode.elm, oldCh = oldVnode.children, ch = vnode.children;
-   if (oldVnode === vnode) return;
+   if (oldVnode === vnode || (isDef(i = vnode.equals) && i(oldVnode, vnode))) return;
    ...
``` 

Now it would be possible to clone the existing vtree (thus avoid invalid patching behaviour) **and** preserve the identity check optimization:
```js
// in my library code
const hh = require("snabbdom/h")
const h = (...args) => {
   const vnode = hh(...args)
   vnode.data.identity = {}
   vnode.equals = (old, vnode) => old.data.identity === vnode.data.identity 
   return vnode
}
```

What do you think? Does this modification make any sense? I already have a forked brach with the modification and tests. If this is ok for you, then I'll make a PR for this.

Thanks!

**Comment by paldepind** [comment]
> However, this disables a very important performance optimization: vnode identity check.

This optimization only applies when using thunks or something similar. Are you using thunks?

**Comment by milankinen** [comment]
>This optimization only applies when using thunks or something similar. Are you using thunks?

Nope. It actually applies to normal `vnode`s as well and when using snabbdom with Observables it's quite easy to take advance of that optimization when using e.g. `combineLatest`. 

You can verify this by debugging my example above (place `debugger` statement between first and second patch: the first `li` vnode is caught by the identity check, thus it and its children are not diffed/patched further).

**Comment by mightyiam** [comment]
There are 7 likes on this issue. But I'm not sure whether they are users who came here because they reused VNodes (which I'm pretty sure is not supported) and got undesired behavior or they support the suggestion of a custom VNode equality check API. Regardless, I've changed the title to make it obvious what this is about.

I am skeptical about the necessity of this feature. I'd like to see evidence of insufficient performance myself.
